doc/designs/quic-design/congestion-control.md

```diff
@@ -0,0 +1,229 @@
+accessing the `ccdata` which should not be a limitation as the calls
+into the module will be done from the packetizer layer of the QUIC
+connection handling and there the eventual writes from multiple threads
+handling individual streams of the connection have to be synchronized
+already to create the packets.
+
+Congestion event handling
+-------------------------
+
+The congestion state of a connection can change only after some event
+happens - i.e., a packet is considered lost meaning the `on_data_lost()`
+is called, or an ack arrives for a packet causing calls to
+`on_data_acked()` or `on_spurious_congestion_event()` functions.
+The congestion control does not produce any timer events by itself.
+
+Exemptions
+----------
+To facilitate probing and to avoid having to always special-case
+probing packets when considering congestion on sending, the
+`set_exemption()` function allows setting number of packets that
+are allowed to be sent overriding the eventual congestion state.
+
+Appendix A
+----------
+
+Proposed header file with comments explaining the individual
+functions. The API is meant to be internal initially so the method
+accessors to set the individual functions will be added later once
+the API is public. Alternatively this might be also implemented
+as fetchable dispatch API.
+
+```C
+/*
+ * Copyright 2022 The OpenSSL Project Authors. All Rights Reserved.
+ *
+ * Licensed under the Apache License 2.0 (the "License").  You may not use
+ * this file except in compliance with the License.  You can obtain a copy
+ * in the file LICENSE in the source distribution or at
+ * https://www.openssl.org/source/license.html
+ */
+
+#include <openssl/ssl.h>
+
+typedef struct ossl_cc_method_st OSSL_CC_METHOD;
+/*
+ * An OSSL_CC_DATA is an externally defined opaque pointer holding
+ * the internal data of the congestion control method
+ */
+typedef struct ossl_cc_data_st OSSL_CC_DATA;
+
+/*
+ * Once this becomes public API then we will need functions to create and
+ * free an OSSL_CC_METHOD, as well as functions to get/set the various
+ * function pointers....unless we make it fetchable.
+ */
+struct ossl_cc_method_st {
+    /*
+     * Create a new OSSL_CC_DATA object to handle the congestion control
+     * calculations.
+     *
+     * |settings| are mandatory settings that will cause the
+     * new() call to fail if they are not understood).
+     * |options| are optional settings that will not
+     * cause the new() call to fail if they are not understood.
+     * |changeables| contain additional parameters that the congestion
+     * control algorithms need that can be updated during the
+     * connection lifetime - for example size of the datagram payload.
+     * To avoid calling a function with OSSL_PARAM array every time
+     * these parameters are changed the addresses of these param
+     * values are considered permanent and the values can be updated
+     * any time.
```

Comment:
I'm not sure this is a good idea. I assume that they need to create such an OSSL_PARAM for each OSSL_CC_DATA, so it's multithread safe. But I'm not sure you can update all OSSL_PARAMs without changing the pointer.

Reply:
> I'm not sure this is a good idea. I assume that they need to create such an OSSL_PARAM for each OSSL_CC_DATA, so it's multithread safe. But I'm not sure you can update all OSSL_PARAMs without changing the pointer.

Please understand that these params would be read-only from the CC implementation point of view. They would be changed only by the calling code. And as these values are per-connection (or per-path in future once we want to support multiple simultaneous paths) they are inherently one-to-one with OSSL_CC_DATA.